### PR TITLE
CASSANDRA-17332 Add support for vnodes in jvm-dtest

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java]: in the dtest api we have

```
if (tokenSupplier == null)
             tokenSupplier = evenlyDistributedTokens(nodeCount, tokenCount);
```

so we only care about token count IFF token supplier isn't provided... so only thing I can think of is to also include

`getTokenSupplier().tokens(1).size() == 1` then not vnodes

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/GossipTest.java]: more changes are needed cross other classes, but its hard to know what without CI

File: build.xml
Comment (jmckenzie-dev) [inline build.xml]: Since this won't build without .13 anyway, should this be 0.0.13?

Comment (dcapwell) [reply build.xml]: sure

File: .build/build-resolver.xml
Comment (dcapwell) [inline .build/build-resolver.xml]: will comment out before merge, will also need to wait for the next jvm-dtest release as well

File: build.xml
Comment (dcapwell) [inline build.xml:81]: this is to allow setting test jvm args from `ant`. MCK didn't like the idea of using environment variables to set this, and ant doesn't propagate by default, so this lets us to define in CI and have ant pass to the test JVM

File: build.xml
Comment (dcapwell) [inline build.xml]: will be 0.0.13 before merge

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:289]: used to be `ClusterUtils.getLocalToken` but that is now `Collection<String>` so can't use it without updating the code to handle it... so simplify to match the old test.

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java]: nit: is "supported" the right word here? It's about whether it's configured or not right?

**Review by jmckenzie-dev — APPROVED** [review]

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java]: if the test explicitly says not to run when vnodes is enabled, then that configuration isn't supported IMO, `isAllowVnodes()` is based 100% on the test disabling vnode support or not

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java:88]: What do we want to do about these TODO's? Create a follow up JIRA to tend to them, take care of them here, leave them for a future dev to work on?

File: test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java:222]: Consider leaving a breadcrumb comment as to why we're leaving this here as a no-op after removing its body (precedent for this in our various ALTER/SELECT/etcStatement.java files in validate())

File: test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java:80]: Have we considered using public static constants for this, or are they ingrained enough in our collective culture that we're fine just using these param names raw in code + config?

We also reference num_tokens raw in ClusterUtils on getTokenCount (and a variety of other places). Given these are pretty calcified by virtue of being in the .yaml and thus part of our public API contract, probably fine just leaving things. Was just an observation while reading through this (lots of raw strings for params floating around here)

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:50]: nit: import ordering

File: test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java:219]: For the tests where we're explicitly configuring our clusters without vnode support (and will thus be skipped on a vnode enabled run), could we document _why_ we run them without vnodes?

File: test/distributed/org/apache/cassandra/distributed/test/PreviewRepairTest.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/test/PreviewRepairTest.java]: nit: import ordering. Have some o.a.c. above and below org.junit here.

File: build.xml
Comment (jmckenzie-dev) [reply build.xml:81]: Could we use a different name here to reflect that? This gives us "test.jvm.args" and "test-jvmargs" which is pretty confusing. Maybe something like "ant.jvm.args" or something?

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java]: nit: capitalization inconsistency. `noVnodeButTokensDoNotMatch`

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (jmckenzie-dev) [reply test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:289]: Yeah, was wondering if this was more general purpose useful (i.e. should live in some kind of Utils class) rather than restricting it to Gossip only.

Might make sense to move this over adjacent to `ClusterUtils.getLocalTokens` as a general purpose helper function even though we only use it in this one place atm. Otherwise we're more likely to duplicate this logic over there in the future when someone needs it if they don't dig enough to find this.

File: test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java
Comment (jmckenzie-dev) [inline test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: Why does this test need num_tokens set to 1 like this? Seems like this is the cause of needing all the song and dance around capturing the skipMessage etc. Also seems somewhat unique to this test, but I can't think of any reason replica filtering as a mechanism wouldn't need to also be tested in a vnode enabled environment? (caveat: haven't read through all the tests here in detail; spent time reviewing the logic around capturing the skip reasoning etc)

File: test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java:80]: `public static constants for this` do you mean `num_tokens`?

We could create a constants class, but rather not in this patch as this code base uses magic constants a lot already.

File: test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java:219]: nope, not the author so no idea why the test failed =D.  Honestly, I feel the test prob made assumptions, so without diving deeper into what those are it was easier for me to disable vnodes (Paxos 100% should support vnode...)

File: test/distributed/org/apache/cassandra/distributed/test/PreviewRepairTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/PreviewRepairTest.java]: we really need check style to yell

File: test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java
Comment (jmckenzie-dev) [reply test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java:80]: Definitely. Just wanted to check my understanding; probably worth making a note to make a follow up ticket to try and eliminate some of these magic constants here.

File: test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java
Comment (jmckenzie-dev) [reply test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java:219]: Hah! Hmm.... maybe a note to just say "TODO: fails with vnode enabled" or something? For a lot of these I can't think of a high level reason why they'd fail (i.e. the functionality they're testing should work in either environment)

File: build.xml
Comment (dcapwell) [reply build.xml:81]: all these settings are for `ant`, so I rather not use `ant.jvm.args` as that I feel makes things more confusing (ant is a java project, so these sound to change the JVM args of ant itself).

`test-jvmargs` is an implementation detail where its trying to have different defaults for j8 and j11; the names are similar true, but I rather rename `test-jvmargs` than `test.jvm.args` as this one is meant to be used from CI

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:289]: so, if we code against 1 token then the test can't really support vnode, so find it best to not generally support a single token API for that reason

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java:88]: so I can't take care of here as its a API/javac problem... since j8 isn't smart enough we have to tell it which method we are targeting (wasn't j11 better here?)

```
public B withTokenSupplier(TokenSupplier tokenSupplier)
public B withTokenSupplier(AbstractBuilder.SingleTokenSupplier tokenSupplier)
```

this was to make the code source compatible but looks like javac defaults to AbstractBuilder.SingleTokenSupplier... only way to remove the cast is to rename `public B withTokenSupplier(TokenSupplier tokenSupplier)`...

Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java:88]: for now I think best to leave to future work, as this cast is only done in `AbstractClusterTest` and "just works" in real test

> withTokenSupplier(TokenSupplier.evenlyDistributedTokens(2))

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:50]: I asked IntelliJ to fix and it didn't do anything... what is the expected order?

There is a block above with junit, do you want this there?  This class doesn't look consistent =(

File: test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: > Why does this test need num_tokens set to 1 like this?

not author so can't justify; my guess is "I want to be sure we don't use vnode", @maedhroz mind commenting?

> Seems like this is the cause of needing all the song and dance around capturing the skipMessage etc

actually no... this is a junit issue -_-. If you rely on static clusters (tests reusing the same cluster) then throwing a skip error will not skip the tests, but will mark the class as failing; aka red build.

Since this class does not allow vnode, the assumption error is thrown in the static method, but we need to skip the tests, so each test needs to do the check... We could change the behavior to allow the cluster to startup with vnode, but have the tests skip if the cluster is vnode... we waste resources is the only side effect (so most of the patch stays the same, just remove the try/catch block)

File: build.xml
Comment (jmckenzie-dev) [reply build.xml:81]: I'm good w/renaming test-jvmargs.

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (jmckenzie-dev) [reply test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:50]: My bad. Thought this was org.junit, didn't parse that it was org.assertj. This is fine.

File: test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java
Comment (jmckenzie-dev) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: Ah, got it. I'd rather just leave it as is and probably put some of what you stated here in a comment in the class (note about cluster reuse and consequences, assumption in static method, etc)

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java
Comment (jmckenzie-dev) [reply test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java:88]: Agreed. Just searched the codebase for TODO; didn't realize how many of these we have littered throughout the codebase.

File: test/distributed/org/apache/cassandra/distributed/test/GossipTest.java
Comment (jmckenzie-dev) [reply test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:289]: Hm. Good point. Promoting this would effectively encourage that behavior by virtue of making it lower friction to engage in.

WFM. Carry on.

File: test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java
Comment (maedhroz) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: I think this was just a case of me explicitly guaranteeing simpler system state for the test when I wrote it. There are only two nodes in this cluster, and it should work just fine with or without vnodes.

Comment (maedhroz) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: Ah, ok, I think the other reason is that in Python dtest space, `ReplicaSideFiltering` does test w/ vnodes.

Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: >  put some of what you stated here in a comment in the class 

so the comment at the catch, just moved to class level?

Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: pushed bigger doc at the skipMessage field explaining why its a thing

File: build.xml
Comment (dcapwell) [reply build.xml:81]: went with `_std-test-jvmargs`

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java:88]:
```
$ grep -r TODO src/ test/ | wc -l
     339
```

yeeeeaaahhhh.....

File: test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: I reverted the skip logic... I remember it was causing issues but it no longer is and CASTest is the same pattern... the junit xml shows the class as skipped...

File: test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/PaxosRepairTest.java:219]: Added the todo and removed `num_tokens` from ReplicaFilteringProtectionTest so it runs with vnodes (it passes locally)

File: test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java]: I removed `num_tokens` and test passes... let the vnodes begin!

File: test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/distributed/impl/AbstractClusterTest.java:102]: I'm not sure there's an easy way out, really: we can remove the SingleTokenSupplier interface to disambiguate though.

File: test/distributed/org/apache/cassandra/distributed/test/CASTest.java
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/distributed/test/CASTest.java:82]: Should we file a jira for this? I can't see anything in this test that should make it fail with vnodes, could be a bug or test bug?

**Review by ifesdjeen — APPROVED** [review]

**Pull request closed without merging**
